<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Pseudotime | Prénom Nom </title> <meta name="author" content="Prénom Nom"> <meta name="description" content="A simple, whitespace theme for academics. Based on [*folio](https://github.com/bogoli/-folio) design. "> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="/assets/img/RightRlogo.png?9419800ae39848847532570d4ca43f3a"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="/projects/sc_pseudotime.html"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Prénom</span> Nom </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">Contact </a> </li> <li class="nav-item "> <a class="nav-link" href="/bioinformatic-analyses/">Bioinformatic analyses </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">Publications </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link"><i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Pseudotime</h1> <p class="post-description"></p> </header> <article> <p>Pseudotime inference, or differentiation trajectory analysis in scRNA-seq, is a computational method used to model the continuous progression of cells as they differentiate from one state to another, such as from stem cells to specialized cell types. By analyzing transcriptomic similarities and differences among individual cells, this approach reconstructs trajectories that are typically visualized on dimensionality reduction spaces like PCA or UMAP, providing an abstract but informative representation of cellular transitions over inferred pseudotime. It uncovers the molecular steps and regulatory programs guiding cell fate decisions, helping identify key transition states, branching points (where trajectories separate), and the genes driving these processes.</p> <p><br></p> <h3>slingshot</h3> <p>Among the <a href="https://github.com/agitter/single-cell-pseudotime" rel="external nofollow noopener" target="_blank">many methods</a> available to model differentiation trajectories on single-cell RNA-seq data, <a href="https://github.com/kstreet13/slingshot" rel="external nofollow noopener" target="_blank">slingshot</a>, and in particular its extension <a href="https://github.com/HectorRDB/condiments" rel="external nofollow noopener" target="_blank">condiments</a>, which can compute separate trajectories for different conditions in the same merged object, stand out. Unlike most other methods, they elegantly build smoothed curves on UMAP coordinates (figure 1A), which are visually more intuitive than tree-like structures (e.g., Minimum Spanning Tree (MST)-based approaches, like <a href="https://github.com/Winnie09/Lamian" rel="external nofollow noopener" target="_blank">Lamian</a>) (figure 1B), and avoids the overfitting tendencies observed in tools like <a href="https://cole-trapnell-lab.github.io/monocle3/" rel="external nofollow noopener" target="_blank">Monocle3</a> (figure 1C).</p> <div class="row"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/sc_pseudotime_1A-480.webp 480w,/assets/img/sc_pseudotime_1A-800.webp 800w,/assets/img/sc_pseudotime_1A-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/img/sc_pseudotime_1A.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="slingshot" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/sc_pseudotime_1B-480.webp 480w,/assets/img/sc_pseudotime_1B-800.webp 800w,/assets/img/sc_pseudotime_1B-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/img/sc_pseudotime_1B.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="Lamian" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/sc_pseudotime_1C-480.webp 480w,/assets/img/sc_pseudotime_1C-800.webp 800w,/assets/img/sc_pseudotime_1C-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/img/sc_pseudotime_1C.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="Monocle3" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Figure 1: Examples of pseudotime analysis on classical monocyte differentiation to non-classical monocytes or macrophages from <a class="citation" href="#Varin2025.04.21.25324260">(Varin* et al., 2025)</a> with three methods. A: Trajectories obtained from <a href="https://github.com/kstreet13/slingshot" rel="external nofollow noopener" target="_blank">slingshot</a>. B: MST structure obtained from <a href="https://github.com/Winnie09/Lamian" rel="external nofollow noopener" target="_blank">Lamian</a>. C: Trajectories obtained from <a href="https://cole-trapnell-lab.github.io/monocle3/" rel="external nofollow noopener" target="_blank">Monocle3</a>. </div> <p><br></p> <h3>tradeSeq</h3> <p>Determining which genes are differentially expressed between trajectories is key to understanding the biological switches responsible for cell fate decisions. For this purpose, the same team behind slingshot developed <a href="https://github.com/statOmics/tradeSeq" rel="external nofollow noopener" target="_blank">tradeSeq</a>, a package built as a direct follow-up to slingshot. TradeSeq fits a negative binomial generalized additive model (NB-GAM) to smooth gene expression along trajectories and provides statistical tools to identify differentially expressed genes between lineages. While tradeSeq includes visualization functions, the package <a href="https://Alexis-varin.github.io/RightOmicsTools" rel="external nofollow noopener" target="_blank">RightOmicsTools</a> further extends these capabilities, offering additional features such as a heatmap with cell density along trajectories (Figure 2A) and an enhanced version of tradeSeq’s scatterplot and smoothed curves (Figure 2B).</p> <div class="row"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/sc_pseudotime_2A-480.webp 480w,/assets/img/sc_pseudotime_2A-800.webp 800w,/assets/img/sc_pseudotime_2A-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/img/sc_pseudotime_2A.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="RightOmicsTools tradeSeq heatmap" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/sc_pseudotime_2B-480.webp 480w,/assets/img/sc_pseudotime_2B-800.webp 800w,/assets/img/sc_pseudotime_2B-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/img/sc_pseudotime_2B.jpg" class="img-fluid rounded z-depth-1" width="100%" height="auto" title="RightOmicsTools tradeSeq scatterplot" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Figure 2: Examples of different visualizations of smoothed gene expression from <a href="https://github.com/statOmics/tradeSeq" rel="external nofollow noopener" target="_blank">tradeSeq</a> obtained using the <a href="https://Alexis-varin.github.io/RightOmicsTools" rel="external nofollow noopener" target="_blank">RightOmicsTools</a> package. A: Heatmap on a single trajectory. B: Scatterplot of normalized expression and average smoothed expression on all trajectories. </div> <p>Finally, you can find additional examples of slingshot and tradeSeq use in <a class="citation" href="#Ramon2025-bm">(Ramon et al., 2025)</a>. <br> <br> <br></p> </article> <h2>Related publications</h2> <div class="publications"> <h2 class="bibliography">2025</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100" style="background-color:#590855;color:#ffffff;"> <div style="color:#ffffff;">Preprint</div> </abbr> <figure> <picture> <img src="/assets/img/publication_preview/Varin_2025_preview.png" class="preview z-depth-1 rounded" width="100%" height="auto" alt="Varin_2025_preview.png" data-zoomable loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div id="Varin2025.04.21.25324260" class="col-sm-8"> <div class="title">A subset of pro-inflammatory CXCL10+ LILRB2+ macrophages derives from recipient monocytes and drives renal allograft rejection</div> <div class="author"> Alexis Varin<sup>*</sup>, Jovanne Palvair<sup>*</sup>, Lennie Messager<sup>*</sup>, Jamal Bamoulid, Yacine Benchikh, Jasper Callemeyn, Mélanie Chaintreuil, Ludivine Dal Zuffo, Didier Ducloux, Imane Farhat, Mathieu Legendre, Laurent Martin, Florian Renosi, Xavier Roussel, Thibaut Vaulet, Maarten Naesens, Claire Tinel, and Baptiste Lamarthée <i class="fa-solid fa-circle-info ml-1" data-toggle="popover" data-placement="top" data-html="true" data-content="* These authors contributed equally"> </i> </div> <div class="periodical"> <em>medRxiv</em>, Apr 2025 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abstract</a> <a href="https://doi.org/10.1101/2025.04.21.25324260" class="doi btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> <a href="https://github.com/InsermRightLab/LILRB2-Varin-et-al" class="code btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="badges"> <span class="altmetric-embed" data-hide-no-mentions="true" data-badge-type="2" data-badge-popover="right" data-doi="10.1101/2025.04.21.25324260"></span> </div> <div class="abstract hidden"> <p>In solid organ transplantation, monocytes and macrophages play a cross-cutting role in the rejection process, irrespective of the transplanted tissue and the type of rejection. Here, we integrated multiple single-cell assays (&gt;150,000 cells) with a broad spectrum of blood-derived and renal allograft-derived cells. We observed 6 myeloid cell trajectories enriched in the allograft during rejection, ranging from circulating CD14+ monocytes to differentiated macrophages in the kidney, with one trajectory culminating in a pro-inflammatory macrophage expressing CXCL9 and CXCL10. By analyzing over 850 biopsies using deconvolution, we report that they are absent in pre-transplant allografts, while these CXCL10+ macrophages are the immune cells most associated with inflammation during acute rejection. Furthermore, a survival study of over 500 biopsies indicates that they increase the risk of graft loss independently of other immune cells. CXCL10+ macrophages differentiate from recipient monocytes, and we have identified 6 major genes associated with their differentiation, including LILRB2. In vitro, mimicking allogenic activation of blood monocytes via the CD47/SIRP-a axis induced overexpression of LILRB2, suggesting that CXCL10+ macrophages are activated by this pathway. Finally, we show that macrophages overexpressing LILRB2 induce the proliferation of autologous T lymphocytes. Altogether, the present study provides further insight into the pro-inflammatory axes of recipient-derived monocytes/macrophages, and suggests LILRB2 as a therapeutic target.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100" style="background-color:#ff8800;color:#212529;"> <div style="color:#212529;">Impact Factor &lt; 5</div> </abbr> <figure> <picture> <img src="/assets/img/publication_preview/Ramon_2025_preview.png" class="preview z-depth-1 rounded" width="100%" height="auto" alt="Ramon_2025_preview.png" data-zoomable loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div id="Ramon2025-bm" class="col-sm-8"> <div class="title">Mature CD209+CD83+CCR7+ dendritic cells infiltrate the arterial wall in giant cell arteritis and derive from in-situ monocyte differentiation</div> <div class="author"> André Ramon, Hélène Greigert, Baptiste Lamarthée, Corentin Richard, Alexis Varin, Claudie Cladière, Coraline Genet, Marion Ciudad, Noémie Klopfenstein, Roman Praliaud, Guillaume Brenac, Georges Tarris, Laurent Martin, Louis Arnould, Pierre-Henry Gabrielle, Catherine Creuzot Garcher, Paul Ornetti, Sylvain Audia, Romain Boidot, Jean-Francis Maillefert, Bernard Bonnotte, and Maxime Samson </div> <div class="periodical"> <em>Sci Rep</em>, Oct 2025 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abstract</a> <a href="https://pubmed.ncbi.nlm.nih.gov/41107388" class="pubmed btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Pubmed</a> <a href="https://doi.org/10.1038/s41598-025-20436-5" class="doi btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> </div> <div class="badges"> <span class="__dimensions_badge_embed__" data-pmid="41107388" data-style="small_rectangle" data-legend="hover-right" style="margin-bottom: 3px;"></span> <span class="altmetric-embed" data-hide-no-mentions="true" data-badge-type="2" data-badge-popover="right" data-pmid="41107388"></span> </div> <div class="abstract hidden"> <p>This study aimed to characterize arterial dendritic cells (DCs) in polymyalgia rheumatica (PMR) and giant cell arteritis (GCA). Bulk RNA-sequencing, RT-PCR and immunofluorescence analyses were performed from temporal arteries from GCA, PMR and control patients. Public single-cell RNA-seq (scRNA-seq) data on Peripheral Blood Mononuclear Cells (PBMCs) were analyzed from three GCA and three control patients. Bulk RNA-Seq and RT-PCR analyses demonstrated a high level of expression of DC lineage markers (CD209), DC maturation markers (CD83, CCR7) and chemokines associated with DC maturation in GCA arteries. The level of expression of DC lineage and DC maturation associated genes was significantly lower in PMR than in GCA arteries and similar between PMR and control arteries. GCA arteries expressed high levels of GM-CSF and IFNG mRNA. ScRNA-seq analysis of GCA PBMCs demonstrated high expression of IFNGR and CSF2R by classical monocytes and cultures of CD14(+) monocytes with GM-CSF and IFN-γwere able to promote their differentiation into monocyte derived-DCs (mo-DCs). This work provides evidence that mo-DCs infiltrate GCA lesions and could be generated under the influence of GM-CSF and IFN-γfrom monocytes infiltrating the arterial wall. Mo-DCs could play an important role in GCA pathogenesis and be targeted by GM-CSF and/or IFN-γinhibitors.</p> </div> </div> </div> </li> </ol> </div> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Prénom Nom. Powered by <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>